keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1615 @@
+
+<!--
+Even if applying deprecation policies, they may still surprise some users.
+-->
+
+### Monitoring Requirements
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+-->
+
+###### How can an operator determine if the feature is in use by workloads?
+
+The operator is the one who would enable the feature, and they would
+know it is in use by looking at the kube-proxy configuration.
+
+###### How can someone using this feature know that it is working for their instance?
+
+- [X] Other (treat as last resort)
+  - Details: If Services still work then the feature is working
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+<!--
+This is your opportunity to define what "normal" quality of service looks like
+for a feature.
+
+It's impossible to provide comprehensive guidance, but at the very
+high level (needs more precise definitions) those may be things like:
+  - per-day percentage of API calls finishing with 5XX errors <= 1%
+  - 99% percentile over day of absolute value from (job creation time minus expected
+    job creation time) for cron job <= 10%
+  - 99.9% of /health requests per day finish with 200 code
+
+These goals will help you determine what you need to measure (SLIs) in the next
+question.
+-->
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+TBD. It's not clear yet what sort of metrics will be interesting. For
+example, in the `iptables` backend we have
+`sync_proxy_rules_iptables_total`, which tells you the total number of
+iptables rules kube-proxy has programmed. But the equivalent metric in
+the `nftables` backend is not going to be as interesting, because many
+of the things that are done with rules in the `iptables` backend will
+be done with maps and sets in the `nftables` backend. Likewise, just
+tallying "total number of rules and set/map elements" is not likely to
+be useful, because the entire point of sets and maps is that they have
+more-or-less **O(1)** behavior, so knowing the number of elements is
+not going to give you much information about how well the system is
+likely to be performing.
+
+- [X] Metrics
+  - Metric names:
+      - ...
+  - Components exposing the metric:
+      - kube-proxy
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+<!--
+Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
+implementation difficulties, etc.).
+-->
+
+### Dependencies
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### Does this feature depend on any specific services running in the cluster?
+
+It may require a newer kernel than some current users have. It does
```

Comment:
Can you explicitly define the min kernel level? You did mention that earlier in the doc, iirc.

Reply:
The examples in the KEP require kernel 5.6, but I discovered later that they require an `nft` binary which is more recent than that. I don't have a good sense of exactly what my current PoC requires (other than that it works on Fedora 36 but requires a few tweaks on RHEL 8).

So...

First, we need to figure out / decide what the required kernel and userspace versions will be. (My theory here is "the lowest common denominator of RHEL 8, and whatever version of Debian the official Kubernetes images are built against".)

Then we need to:
1. Make sure this makes it into the end user documentation.
2. Check this at runtime (with `nft --check` as discussed earlier)
3. Have clear internal documentation (and maybe assertions) to make sure people don't accidentally bump these requirements in the future.

Reply:
That's fine - this is not required for Alpha and we will revisit it for Beta once we already will have a working version and will know much more about it.

Reply:
related https://github.com/kubernetes/kubernetes/issues/116799